### ISSUE TITLE
Implement tensor and numerical runtime foundations

### ISSUE THREAD

**Issue by samifouad** [dekaruntime/dsc#19]
## Goal

Build the numerical layer required by data science and AI packages on top of the Deka data-kernel decisions. The implementation should expose readable Deka APIs while allowing Rust-backed native kernels and explicit device backends.

## Required concepts

A tensor has storage, dtype, shape, strides, and device/location.

Shapes generalize scalars, vectors, matrices, images, and batches:

\`\`\`text
scalar       []
vector       [3]
matrix       [2, 3]
image        [height, width, channels]
text batch   [batch, tokens, embedding_size]
\`\`\`

## Initial operations

- construction, zeros, ones, and ranges
- indexing and slicing without unnecessary copies
- reshape, transpose, and views
- reductions by axis
- broadcasting with explicit compatibility errors
- elementwise arithmetic and comparisons
- matrix multiplication
- softmax, log-softmax, exp, log, and normalization
- stable cross-entropy loss
- deterministic random generation through an explicit RNG value

Invalid shapes, devices, dtypes, and indexing must return typed errors. Never silently flatten, truncate, widen, or copy across devices when that changes observable behavior.

## Native architecture

Keep the Deka-facing operation APIs stable while allowing Rust CPU/SIMD kernels, optional CUDA or Metal backends, a WASM/browser backend where practical, and explicit unsupported-capability errors. Do not make browser parity a prerequisite for native functionality.

## Acceptance criteria

- Native tests cover shape, dtype, stride, view, device, and error semantics.
- Numerical tests compare representative operations against a trusted reference.
- Tests prove no hidden global device or RNG state is required.
- Performance-sensitive operations have a documented Rust/native path.
- The API is sufficient to implement a small transformer without handwritten low-level storage code. -codex
